### PR TITLE
`<type_traits>`: `std::byte` is `_Is_trivially_swappable_v`

### LINKED ISSUE #2039: `<type_traits>`: `std::byte` is not `_Is_trivially_swappable`
**Describe the bug**
`std::byte` is not `_Is_trivially_swappable`, so code generation is worse than for `unsigned char`

This problem is best seen at godbolt: https://godbolt.org/z/qxP8fYd3v

**Expected behavior**
`std::byte` should have same speed as `unsigned char`

**STL version**
    ```
    Microsoft Visual Studio Community 2022 Preview 1.1
    Version 17.0.0
    ```

**Additional context**
 DevCom-1467265 and Microsoft-internal VSO-1352013 / AB#1352013

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2039

### CODE REVIEW

**Review by miscco — APPROVED** [review]

File: tests/std/tests/VSO_0180469_ptr_cat/test.compile.pass.cpp [hunk 1]
```diff
@@ -393,6 +393,8 @@ void equal_safe_test_cases() {
     test_case_Equal_memcmp_is_safe<false, base_class, base_class>();
 
 #ifdef __cpp_lib_byte
+    // GH-2039
+    static_assert(_Is_trivially_swappable_v<byte>);
```
Comment (AdamBucior) [inline tests/std/tests/VSO_0180469_ptr_cat/test.compile.pass.cpp:397]: Should we add some more test for `_Is_trivially_swappable_v` (not only for `byte`)? I couldn't find any.

**Review by AdamBucior — COMMENTED** [review]
I think we should also add the specializations for types like `duration` and `time_point` (if their underlying type is trivially swappable)

**Comment by fsb4000** [comment]
@AdamBucior 
>I think we should also add the specializations for types like duration and time_point

How does such specialization look like? I'm sorry I do not understand.

**Comment by AdamBucior** [comment]
> How does such specialization look like? I'm sorry I do not understand.

```
template <class _Rep, class _Period>
inline constexpr bool _Is_trivially_swappable_v<duration<_Rep, _Period>> = _Is_trivially_swappable_v<_Rep>;
```

File: stl/inc/chrono [hunk 2]
```diff
@@ -278,6 +279,13 @@ namespace chrono {
     };
 } // namespace chrono
 
+template <class _Rep, class _Period>
+_INLINE_VAR constexpr bool _Is_trivially_swappable_v<chrono::duration<_Rep, _Period>> = _Is_trivially_swappable_v<_Rep>;
+
+template <class _Clock, class _Duration>
+_INLINE_VAR constexpr bool _Is_trivially_swappable_v<chrono::time_point<_Clock, _Duration>> =
+    _Is_trivially_swappable_v<_Duration>;
```
Comment (CaseyCarter) [inline stl/inc/chrono:287]: I'm mildly concerned that this relationship might not hold for program-defined specializations of `duration` and `time_point`. Should we use the `_From_primary` technique from `iterator_traits` to constrain this to library-defined specializations only?

File: tests/std/tests/GH_002039_byte_is_not_trivially_swappable/test.compile.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+enum class byte2 : unsigned char {};
+STATIC_ASSERT(_Is_trivially_swappable_v<byte2>);
+
+enum class byte3 : unsigned char {};
+
+void swap(byte3& a, byte3& b) {
+    byte3 temp{a};
+    a = b;
+    b = temp;
+}
```
Comment (CaseyCarter) [inline tests/std/tests/GH_002039_byte_is_not_trivially_swappable/test.compile.pass.cpp:45]: It's conventional to declare but not define functions in COMPILE-ONLY tests - function bodies are just noise. I would
 ```suggestion
void swap(byte3&, byte3&);
```
here, and similarly avoid defining the non-trivial member functions in the classes below.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/chrono (hunk 2 above)
Comment (StephanTLavavej) [reply stl/inc/chrono:287]: I think that this is fine as-is - I can't imagine why a program would ever attempt to specialize `duration` and `time_point` (unlike `hash` or `less`, or even `allocator`), and even if so, how they could be non-trivially-swappable differently from their `_Duration` and `_Rep`. Theoretically it is possible, but I believe we will never ever see it in practice.

The current PR handles the case where `duration` and `time_point` are the STL's primary templates, but a user-defined representation type (possibly a bignum) is being used - that is also very rare but within the realm of possibility that we should attempt to handle.

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/chrono (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/chrono:287]: As long as we're all clear that I get to say "I told you so" if there's ever a bug report, I'm ok with this =)

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for extending this optimization! :rocket: :rocket: :cat2:

**Pull request merged**